### PR TITLE
New triggers implementation with redirect support

### PR DESCRIPTION
**Pull request by arunoda**
This is a re-implementation with our [triggers API](https://github.com/meteorhacks/flow-router/pull/59) with redirect support with a clean codebase.

This is the implementation of the redirect API.

~~~js
FlowRouter.route('/', {
  triggersEnter: [function(context, redirect) {
    redirect('/some-other-path');
  }],
  action: function(_params) {
    throw new Error("this should not get called");
  }
});
~~~

When someone route to `/` then he'll be redirected to `/some-other-path`. In this case, action won't get called or any exit trigger if exists.

This can be done for exit triggers as well. 

We can also do this prevent route changes like this.

~~~js
FlowRouter.route('/', {
  triggersExit: [function(context, redirect) {
   // may be in exit handlers, we should accept to redirect without an URL.
   // If so, it will simply stop. But, this should be not be in the enter triggers
    redirect(context.path);
  }],
});
~~~

In the above case, it is impossible to leave from route `/`. 

Here are some properties of redirect

* redirect must have an URL
* redirect must be called within the same eventloop cycle (no async or inside a Tracker)
* redirect cannot be called multiple times

> All these have been added to make sure you are not doing any reactive code inside the Router.

### CODE REVIEW

**Comment by arunoda** [comment]
@delgermurun check this. Give it a try and help me to figure out any issues?

File: client/router.js [hunk 1]
```diff
@@ -353,99 +373,9 @@ Router.prototype._buildTracker = function() {
...
-          fn(self._current);
-        }
-      });
-
-      next();
-    });
-  }
+  // we need to trigger the above invalidations immediately
+  // otherwise, we need to face some issues with route context swapping
+  Tracker.flush();
```
Comment (arunoda) [inline client/router.js]: Check how flush behave if a user change the route from a template helper. 

That's weird. But possible.

**Comment by delgermurun** [comment]
@arunoda Nice refactoring! Working good. What's the issues :)?

**Comment by arunoda** [comment]
@delgermurun I mean if there are any issues :)

**Comment by arunoda** [comment]
Everything looks good. I think we can do a release.

**Comment by delgermurun** [comment]
Yeah. Do it.

**Comment by cstrat** [comment]
This is perfect, thanks Aronuda :+1:

**Comment by derwaldgeist** [comment]
I'm trying to redirect to a sub-route of the current route, for instance from /foo to /foo/bar. This works (shows the right template for the router), but the URL ifself is not updated (still shows /foo). The same applies to my navigation bar which is testing for the route to highlight the right navigation elements. If I'm using FlowRouter.go() instead of redirect(), the same happens. How can I update the URL and my navigation template once a redirect happens?

**Comment by jjman505** [comment]
How do we use the FlowRouter.withReplaceState(fn) with this redirect API? I want to redirect, but only keep the last redirected-to route in the history.

**Comment by tangweikun** [comment]
飞

**Pull request merged**